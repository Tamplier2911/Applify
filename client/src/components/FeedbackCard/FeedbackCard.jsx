// import "./FeedbackCard.scss";
import React from "react";

// redux
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { selectCurrentLanguage } from "../../redux/lang/lang.selectors";
import { deleteFeedbackStart } from "../../redux/feedbacks/feedbacks.actions";

// components
import Button from "../Button/Button";

// data formaters
import getImageRelativePath from "../../utils/PathTransformations/getImageRelativePath";
import transformDateToLocaleString from "../../utils/DataTransformations/transformDateToLocaleString";

// JS Rendering CSS
import {
  FeedbackCardContainer,
  FeedbackCardHeader,
  FeedbackCardHeaderName,
  FeedbackCardBody,
  FeedbackCardBodyImgWrapper,
  FeedbackCardBodyImage,
  FeedbackCarBodyRating,
  FeedbackCarBodySVG,
  FeedbackCarBodyFeedback,
  FeedbackCardBodyContent,
  FeedbackCarBodyQuoStart,
  FeedbackCarBodyQueEnd,
  FeedbackCarBodyDate,
  FeedbackCarControlls,
  FeedbackCarControllsLink,
} from "./FeedbackCardStyles";

// component constants
import feedbackCardData from "./FeedbackCardConstants";

const FeedbackCard = ({ data, deleteFeedbackStart, lang }) => {
  const { rating, createdAt, _id, feedback, user, color } = data ? data : {};
  const long = feedback.slice(0, 80)[79];
  const shortenFeedback = feedback.slice(0, 80);
  const image = getImageRelativePath(user ? user.photo : "");
  const date = transformDateToLocaleString(createdAt);
  const { feedbackCardDelete, feedbackCardUpdate } = feedbackCardData[lang];
  return (
    <FeedbackCardContainer key={_id}>
      <FeedbackCardHeader>
        <FeedbackCardHeaderName>{user.name}</FeedbackCardHeaderName>
      </FeedbackCardHeader>
      <FeedbackCardBody>
        <FeedbackCardBodyImgWrapper color={color}>
          <FeedbackCardBodyImage src={`${image}`} alt="happy user" />
        </FeedbackCardBodyImgWrapper>
        <FeedbackCarBodyRating>
          {Array.from(new Array(Number(rating)), (el, i) => i + 1).map(
            (el, i) => {
              return <FeedbackCarBodySVG key={i} />;
            }
          )}
        </FeedbackCarBodyRating>
        <FeedbackCarBodyFeedback>
          <FeedbackCarBodyQuoStart color={color}>
            &ldquo;
          </FeedbackCarBodyQuoStart>
          <FeedbackCardBodyContent>
            {long ? shortenFeedback + "..." : shortenFeedback}
          </FeedbackCardBodyContent>
          <FeedbackCarBodyQueEnd color={color}>&bdquo;</FeedbackCarBodyQueEnd>
        </FeedbackCarBodyFeedback>
        <FeedbackCarBodyDate>{date}</FeedbackCarBodyDate>
      </FeedbackCardBody>
      <FeedbackCarControlls>
        <Button
          type="button"
          value={feedbackCardDelete}
          click={() => deleteFeedbackStart(_id)}
        ></Button>
        <FeedbackCarControllsLink to={`/profile/feeds/${_id}`}>
          {feedbackCardUpdate}
        </FeedbackCarControllsLink>
      </FeedbackCarControlls>
    </FeedbackCardContainer>
  );
};

const mapStateToProps = createStructuredSelector({
  lang: selectCurrentLanguage,
});

export default connect(mapStateToProps, { deleteFeedbackStart })(FeedbackCard);
