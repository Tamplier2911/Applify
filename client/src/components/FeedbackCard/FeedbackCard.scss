/*

.card {
  display: grid;
  grid-template-rows: 4fr 6fr 1fr;
  box-shadow: 0rem 0rem 0.8rem var(--cl-font);
  padding: 1rem;
  transition: box-shadow 0.3s;
  font-family: var(--font-title);

  &:hover {
    box-shadow: 0rem 0rem 0.8rem var(--cl-primary);
  }

  &__header {
    display: grid;
    // align-content: center;

    &--name {
      text-align: center;
      font-size: 1.8rem;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }
    &--email {
    }
  }
  &__body {
    position: relative;
    display: grid;
    // grid-auto-rows: min-content;
    grid-template-rows: min-content min-content 1fr;

    &--wrapper {
      position: absolute;
      width: 10rem;
      height: 10rem;
      overflow: hidden;
      border-radius: 50%;
      left: 50%;
      top: -10rem;
      transform: translate(-50%);
    }
    &--img {
      width: 100%;
      height: 100%;
      display: block;
      object-fit: cover;
    }

    &--rating {
      display: flex;
      justify-content: center;
      padding: 2rem 0;

      & svg:not(:last-child) {
        margin-right: 1rem;
      }
    }

    &--svg {
      width: 3rem;
      height: 3rem;
    }

    &--feedback {
      display: grid;
      grid-template-columns: min-content 1fr min-content;

      & div {
        align-self: center;
        padding: 1rem;
        text-align: center;
      }
    }

    &--quoteStart {
      align-self: start;
      display: inline-block;
      font-size: 6rem;
    }

    &--quoteEnd {
      align-self: end;
      display: inline-block;
      font-size: 6rem;
    }

    &--date {
      align-self: center;
      text-align: center;
    }
  }
  &__controlls {
    display: grid;
    grid-template-columns: repeat(2, 1fr);
    grid-column-gap: 1rem;

    &--link {
      display: grid;
      align-items: center;
      text-align: center;
      &:link,
      &:visited {
        text-decoration: none;
        color: var(--cl-white);
        background-color: var(--cl-font);
        transition: background-color 0.3s;
      }
      &:hover,
      &:active {
        background-color: var(--cl-primary);
      }
    }
  }
}

*/
