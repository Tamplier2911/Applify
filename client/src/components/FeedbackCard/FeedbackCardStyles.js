import styled, { css } from "styled-components";
import { Link } from "react-router-dom";
import ScrollAnimation from "react-animate-on-scroll";

// SVG
import { ReactComponent as StartSVG } from "../../assets/svg/star.svg";

const quoteSharedStyles = css`
  display: inline-block;
  font-size: 6rem;
`;

const getColor = (props) => {
  const { color } = props;
  return `color: ${color};`;
};

const getBoxShadow = (props) => {
  const { color } = props;
  return `box-shadow: 0rem 0rem 0.8rem ${color};`;
};

export const FeedbackCardContainer = styled(ScrollAnimation)`
  display: grid;
  grid-template-rows: 4fr 6fr 1fr;
  box-shadow: 0rem 0rem 0.8rem var(--cl-font);
  padding: 1rem;
  transition: box-shadow 0.3s;
  font-family: var(--font-title);

  &:hover {
    box-shadow: 0rem 0rem 0.8rem var(--cl-primary);
  }
`;

export const FeedbackCardHeader = styled.div`
  display: grid;
`;

export const FeedbackCardHeaderName = styled.div`
  text-align: center;
  font-size: 1.8rem;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

export const FeedbackCardBody = styled.div`
  position: relative;
  display: grid;
  grid-template-rows: min-content min-content 1fr;
`;

export const FeedbackCardBodyImgWrapper = styled.div`
  ${getBoxShadow}
  position: absolute;
  width: 10rem;
  height: 10rem;
  overflow: hidden;
  border-radius: 50%;
  left: 50%;
  top: -10rem;
  transform: translate(-50%);
`;

export const FeedbackCardBodyImage = styled.img`
  width: 100%;
  height: 100%;
  display: block;
  object-fit: cover;
`;

export const FeedbackCarBodyRating = styled.div`
  display: flex;
  justify-content: center;
  padding: 2rem 0;

  & svg:not(:last-child) {
    margin-right: 1rem;
  }
`;

export const FeedbackCarBodySVG = styled(StartSVG)`
  width: 3rem;
  height: 3rem;
`;

export const FeedbackCarBodyFeedback = styled.div`
  display: grid;
  grid-template-columns: min-content 1fr min-content;
`;

export const FeedbackCardBodyContent = styled.div`
  align-self: center;
  padding: 1rem;
  text-align: center;
`;

export const FeedbackCarBodyQuoStart = styled.span`
  align-self: start;
  ${quoteSharedStyles}
  ${getColor}
`;

export const FeedbackCarBodyQueEnd = styled.span`
  align-self: end;
  ${quoteSharedStyles}
  ${getColor}
`;

export const FeedbackCarBodyDate = styled.div`
  align-self: center;
  text-align: center;
`;

export const FeedbackCarControlls = styled.div`
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  grid-column-gap: 1rem;
`;

export const FeedbackCarControllsLink = styled(Link)`
  display: grid;
  align-items: center;
  text-align: center;
  &:link,
  &:visited {
    text-decoration: none;
    color: var(--cl-white);
    background-color: var(--cl-font);
    transition: background-color 0.3s;
  }
  &:hover,
  &:active {
    background-color: var(--cl-primary);
  }
`;
