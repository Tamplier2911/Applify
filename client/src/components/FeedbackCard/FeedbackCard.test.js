import React from "react";
import toJson from "enzyme-to-json";

// component
import FeedbackCard from "./FeedbackCard";

// utils
import {
  setShallowC,
  findByTestAttr,
  checkProps,
  storeFactory,
} from "../../tests/testUtils";

describe("<FeedbackCard />", () => {
  const initialStore = {
    lang: {
      languagePanelHidden: true,
      currentLanguage: "eng",
    },
  };
  const store = storeFactory(initialStore);
  const data = {
    color: "#fff",
    createdAt: "2020",
    feedback: "good",
    rating: "5",
    user: {},
  };

  let wrapper;
  beforeEach(() => {
    wrapper = setShallowC(<FeedbackCard store={store} data={data} />);
  });

  it("renders without an error", () => {
    const feedbackCard = findByTestAttr(wrapper, "feedback-card");
    expect(feedbackCard.length).toBe(1);
  });

  it("renders child - card body", () => {
    const feedbackCardBody = findByTestAttr(wrapper, "feedback-card-body");
    expect(feedbackCardBody.length).toBe(1);
  });

  it("renders child - card controlls", () => {
    const feedbackCardControlls = findByTestAttr(
      wrapper,
      "feedback-card-controlls"
    );
    expect(feedbackCardControlls.length).toBe(1);
  });

  it("recieves expected props", () => {
    const expectedProps = {
      color: "#fff",
      createdAt: "2020",
      feedback: "good",
      rating: "5",
      user: {},
    };
    const err = checkProps(FeedbackCard, expectedProps, "props");
    expect(err).toBeUndefined();
  });

  it("matches snapshot", () => {
    const feedbackCard = findByTestAttr(wrapper, "feedback-card");
    expect(toJson(feedbackCard)).toMatchSnapshot();
  });
});
