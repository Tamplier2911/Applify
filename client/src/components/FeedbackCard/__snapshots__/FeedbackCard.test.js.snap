// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FeedbackCard /> matches snapshot 1`] = `ShallowWrapper {}`;
